AntreaNetworkPolicy related tests failed randomly in kind clusters

**Describe the bug**

I have met this kind of failures several times in kind clusters. Normally the same tests can pass if rerunning them.
```
2021-07-13T14:04:39.3057811Z     --- FAIL: TestAntreaPolicy/TestGroupDefaultDENY (73.58s)
2021-07-13T14:04:39.3060028Z         --- FAIL: TestAntreaPolicy/TestGroupDefaultDENY/Case=ACNPAllowXBtoA (17.37s)
2021-07-13T14:04:39.3061680Z         --- PASS: TestAntreaPolicy/TestGroupDefaultDENY/Case=ACNPAllowXBtoYA (18.08s)
2021-07-13T14:04:39.3063329Z         --- PASS: TestAntreaPolicy/TestGroupDefaultDENY/Case=ACNPPriorityOverrideDefaultDeny (17.49s)
```

```
2021-07-13T13:50:57.3006901Z === RUN   TestAntreaPolicy/TestGroupDefaultDENY/Case=ACNPAllowXBtoA
2021-07-13T13:50:57.3008637Z I0713 13:50:57.300523   18647 antreapolicy_test.go:2359] running test case ACNP Allow X/B to A
2021-07-13T13:50:57.3009813Z I0713 13:50:57.300555   18647 antreapolicy_test.go:2361] running step Port 80 of test case ACNP Allow X/B to A
2021-07-13T13:50:57.3011715Z I0713 13:50:57.300566   18647 k8s_util.go:570] Creating/updating ClusterNetworkPolicy acnp-allow-xb-to-a
2021-07-13T13:50:57.4214557Z I0713 13:50:57.421238   18647 util.go:44] Confirming ready status costs 102.429303ms
2021-07-13T13:51:10.5503362Z ##[error]    antreapolicy_test.go:2376: failure -- 1 wrong results
2021-07-13T13:51:10.5509558Z reachability: correct:80, incorrect:1, result=false
2021-07-13T13:51:10.5510373Z 
2021-07-13T13:51:10.5511074Z expected:
2021-07-13T13:51:10.5511293Z 
2021-07-13T13:51:10.5511871Z -	x/a	y/a	z/a	x/b	y/b	z/b	x/c	y/c	z/c
2021-07-13T13:51:10.5512441Z x/a	Con	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5512882Z y/a	Drp	Con	Drp	Drp	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5513327Z z/a	Drp	Drp	Con	Drp	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5513751Z x/b	Con	Con	Con	Con	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5514187Z y/b	Drp	Drp	Drp	Drp	Con	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5514613Z z/b	Drp	Drp	Drp	Drp	Drp	Con	Drp	Drp	Drp
2021-07-13T13:51:10.5515060Z x/c	Drp	Drp	Drp	Drp	Drp	Drp	Con	Drp	Drp
2021-07-13T13:51:10.5515499Z y/c	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Con	Drp
2021-07-13T13:51:10.5515938Z z/c	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Con
2021-07-13T13:51:10.5516194Z 
2021-07-13T13:51:10.5516368Z 
2021-07-13T13:51:10.5516638Z observed:
2021-07-13T13:51:10.5516838Z 
2021-07-13T13:51:10.5517330Z -	x/a	y/a	z/a	x/b	y/b	z/b	x/c	y/c	z/c
2021-07-13T13:51:10.5517738Z x/a	Con	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5518178Z y/a	Drp	Con	Drp	Drp	Drp	Drp	Drp	Drp	Con
2021-07-13T13:51:10.5518599Z z/a	Drp	Drp	Con	Drp	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5519037Z x/b	Con	Con	Con	Con	Drp	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5519459Z y/b	Drp	Drp	Drp	Drp	Con	Drp	Drp	Drp	Drp
2021-07-13T13:51:10.5519898Z z/b	Drp	Drp	Drp	Drp	Drp	Con	Drp	Drp	Drp
2021-07-13T13:51:10.5520320Z x/c	Drp	Drp	Drp	Drp	Drp	Drp	Con	Drp	Drp
2021-07-13T13:51:10.5520757Z y/c	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Con	Drp
2021-07-13T13:51:10.5521198Z z/c	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Drp	Con
2021-07-13T13:51:10.5521635Z 
2021-07-13T13:51:10.5521823Z 
2021-07-13T13:51:10.5522135Z comparison:
2021-07-13T13:51:10.5522355Z 
2021-07-13T13:51:10.5522860Z -	x/a	y/a	z/a	x/b	y/b	z/b	x/c	y/c	z/c
2021-07-13T13:51:10.5523203Z x/a	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5523523Z y/a	.	.	.	.	.	.	.	.	X
2021-07-13T13:51:10.5523840Z z/a	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5524137Z x/b	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5524457Z y/b	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5525075Z z/b	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5525361Z x/c	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5525632Z y/c	.	.	.	.	.	.	.	.	.
2021-07-13T13:51:10.5526109Z z/c	.	.	.	.	.	.	.	.	.
```

**Expected**
The tests should be stable on kind clusters.

**Actual behavior**
It failed randomly.

Logs:
[e2e.log](https://github.com/antrea-io/antrea/files/6810363/e2e.log)
[e2e-kind-encap-no-proxy.tar.gz (1).zip](https://github.com/antrea-io/antrea/files/6810349/e2e-kind-encap-no-proxy.tar.gz.1.zip)

@tnqn I also saw the following [intermittent failure](https://github.com/antrea-io/antrea/pull/2308#issuecomment-871754228) in Antrea network policy tests in Kind clusters along with the one in the above comment. Piggybacking on this issue. Hope that is ok.

```
--- FAIL: TestAntreaPolicy/TestGroupNoK8sNP (742.02s)
    --- PASS: TestAntreaPolicy/TestGroupNoK8sNP/Case=ACNPAllowNoDefaultIsolationTCP (12.83s)
    ........(many pass tests)
    --- FAIL: TestAntreaPolicy/TestGroupNoK8sNP/Case=AppliedToPerRule (29.28s)
```

--- a/multicluster/test/e2e/framework.go
+++ b/multicluster/test/e2e/framework.go
@@ -363,7 +363,10 @@ func (data *TestData) probe(
 	}
 	log.Tracef("Running: kubectl exec %s -c %s -n %s -- %s", podName, containerName, podNamespace, strings.Join(cmd, " "))
 	stdout, stderr, err := data.runCommandFromPod(cluster, podNamespace, podName, containerName, cmd)
-	if err != nil {
+	// It needs to check both err and stderr because:
+	// 1. The probe tried 3 times. If it checks err only, failure+failure+success would be considered connected.
+	// 2. There might be an issue in Pod exec API that it sometimes doesn't return error when the probe fails. See #2394.
+	if err != nil || stderr != "" {
 		// log this error as trace since may be an expected failure
 		log.Tracef("%s -> %s: error when running command: err - %v /// stdout - %s /// stderr - %s", podName, dstName, err, stdout, stderr)
 		// If err != nil and stderr == "", then it means this probe failed because of

--- a/test/e2e/k8s_util.go
+++ b/test/e2e/k8s_util.go
@@ -109,7 +109,10 @@ func (k *KubernetesUtils) probe(
 	}
 	log.Tracef("Running: kubectl exec %s -c %s -n %s -- %s", pod.Name, containerName, pod.Namespace, strings.Join(cmd, " "))
 	stdout, stderr, err := k.runCommandFromPod(pod.Namespace, pod.Name, containerName, cmd)
-	if err != nil {
+	// It needs to check both err and stderr because:
+	// 1. The probe tried 3 times. If it checks err only, failure+failure+success would be considered connected.
+	// 2. There might be an issue in Pod exec API that it sometimes doesn't return error when the probe fails. See #2394.
+	if err != nil || stderr != "" {
 		// log this error as trace since may be an expected failure
 		log.Tracef("%s -> %s: error when running command: err - %v /// stdout - %s /// stderr - %s", podName, dstName, err, stdout, stderr)
 		// If err != nil and stderr == "", then it means this probe failed because of

